sources/rest_api/auth.py

```diff
@@ -0,0 +1,174 @@
+
+    def __call__(self, request: PreparedRequest) -> PreparedRequest:
+        encoded = b64encode(f"{self.username}:{self.password}".encode()).decode()
+        request.headers["Authorization"] = f"Basic {encoded}"
+        return request
+
+
+@configspec
+class OAuth2AuthBase(AuthConfigBase):
+    """Base class for oauth2 authenticators. requires access_token"""
+    # TODO: Separate class for flows (implicit, authorization_code, client_credentials, etc)
+    type: Final[Literal["oauth2"]] = "oauth2"
+    access_token: TSecretStrValue
+
+    def __init__(self, access_token: TSecretStrValue) -> None:
+        self.access_token = access_token
+
+    def __call__(self, request: PreparedRequest) -> PreparedRequest:
+        request.headers["Authorization"] = f"Bearer {self.access_token}"
+        return request
+
+
+@configspec
+class OAuthJWTAuth(BearerTokenAuth):
+    """This is a form of Bearer auth, actually there's not standard way to declare it in openAPI"""
+    format: Final[Literal["JWT"]] = "JWT"
+
+    client_id: str
+    private_key: TSecretStrValue
+    auth_endpoint: str
+    scopes: Optional[str] = None
+    headers: Optional[Dict[str, str]] = None
+    private_key_passphrase: Optional[TSecretStrValue] = None
+
+    def __init__(
+        self,
+        client_id: str,
+        private_key: TSecretStrValue,
+        auth_endpoint: str,
+        scopes: str,
+        headers: Optional[Dict[str, str]] = None,
+        private_key_passphrase: Optional[TSecretStrValue] = None,
+    ):
+        self.client_id = client_id
+        self.private_key = private_key
+        self.private_key_passphrase = private_key_passphrase
+        self.auth_endpoint = auth_endpoint
+        self.scopes = scopes if isinstance(scopes, str) else " ".join(scopes)
+        self.headers = headers
+        self.token = None
+        self.token_expiry: Optional[pendulum.DateTime] = None
+
+    def __call__(self, r: PreparedRequest) -> PreparedRequest:
+        if self.token is None or self.is_token_expired():
+            self.obtain_token()
+        r.headers["Authorization"] = f"Bearer {self.token}"
+        return r
+
+    def is_token_expired(self) -> bool:
+        return not self.token_expiry or pendulum.now() >= self.token_expiry
+
+    def obtain_token(self) -> None:
+        payload = self.create_jwt_payload()
+        data = {
+            "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
+            "assertion": jwt.encode(
+                payload, self.load_private_key(), algorithm="RS256"
+            ),
+        }
+
+        logger.debug(f"Obtaining token from {self.auth_endpoint}")
+
+        response = requests.post(self.auth_endpoint, headers=self.headers, data=data)
+        response.raise_for_status()
+
+        token_response = response.json()
+        self.token = token_response["access_token"]
+        self.token_expiry = pendulum.now().add(
+            seconds=token_response.get("expires_in", 3600)
```

Comment:
should we make default expiry configurable and maybe <1h just in case if token leaks?